pkg/kubelet/lifecycle/handlers.go

```diff
@@ -132,6 +133,17 @@ func (hr *handlerRunner) runHTTPHandler(pod *v1.Pod, container *v1.Container, ha
 		}
 		resp, err := hr.httpDoer.Do(req)
 		discardHTTPRespBody(resp)
+
+		if err != nil {
+			if urlerr, ok := err.(*url.Error); ok && urlerr.Err.Error() == "http: server gave HTTP response to HTTPS client" {
```

Comment:
I'd suggest putting this check in an `isHTTPResponseError(err)` function we can unit test with the error returned from a test server

```go
func isHTTPResponseError(err error) bool {
	if err == nil {
		return false
	}
	urlErr := &url.Error{}
	if !errors.As(err, &urlErr) {
		return false
	}
	return strings.Contains(urlErr.Err.Error(), "server gave HTTP response to HTTPS client")
}
```

tested like this:

```go
s := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {}))
defer s.Close()
req, err := http.NewRequest("GET", s.URL, nil)
if err != nil {
	t.Fatal(err)
}
req.URL.Scheme = "https"
_, err := http.DefaultClient.Do(req)
if !isHTTPResponseError(err) {
	t.Fatalf("expected isHTTPResponseError from %#v", err)
}
```

Reply:
that way the lifecycle unit test exercises the overall fallback logic, and a targeted unit test exercises the stdlib message matching